iseq.c

```diff
@@ -248,230 +245,155 @@ rb_iseq_each_value(const rb_iseq_t *iseq, iseq_value_itr_t * func, void *data)
-                    entry->iseq = (rb_iseq_t *)rb_gc_location((VALUE)entry->iseq);
+                    iseq_scan_bits(i, bits, code, original_iseq);
                 }
             }
         }
-#if USE_MJIT
-        mjit_update_references(iseq);
-#endif
-#if USE_YJIT
-        rb_yjit_iseq_update_references(body->yjit_payload);
-#endif
     }
 }
 
-static VALUE
-each_insn_value(void *ctx, VALUE obj)
-{
-    rb_gc_mark_movable(obj);
-    return obj;
-}
-
 void
-rb_iseq_mark(const rb_iseq_t *iseq)
+rb_iseq_mark_and_update(rb_iseq_t *iseq, bool reference_updating)
 {
     RUBY_MARK_ENTER("iseq");
 
-    RUBY_MARK_UNLESS_NULL(iseq->wrapper);
+    rb_gc_mark_and_move(&iseq->wrapper);
 
     if (ISEQ_BODY(iseq)) {
-        const struct rb_iseq_constant_body *const body = ISEQ_BODY(iseq);
+        struct rb_iseq_constant_body *body = ISEQ_BODY(iseq);
 
-        rb_iseq_each_value(iseq, each_insn_value, NULL);
+        rb_iseq_mark_and_move_each_value(iseq, reference_updating ? ISEQ_ORIGINAL_ISEQ(iseq) : NULL);
 
-        rb_gc_mark_movable(body->variable.coverage);
-        rb_gc_mark_movable(body->variable.pc2branchindex);
-        rb_gc_mark_movable(body->variable.script_lines);
-        rb_gc_mark_movable(body->location.label);
-        rb_gc_mark_movable(body->location.base_label);
-        rb_gc_mark_movable(body->location.pathobj);
-        RUBY_MARK_MOVABLE_UNLESS_NULL((VALUE)body->mandatory_only_iseq);
-        RUBY_MARK_MOVABLE_UNLESS_NULL((VALUE)body->parent_iseq);
+        rb_gc_mark_and_move(&body->variable.coverage);
+        rb_gc_mark_and_move(&body->variable.pc2branchindex);
+        rb_gc_mark_and_move(&body->variable.script_lines);
+        rb_gc_mark_and_move(&body->location.label);
+        rb_gc_mark_and_move(&body->location.base_label);
+        rb_gc_mark_and_move(&body->location.pathobj);
+        if (body->local_iseq) rb_gc_mark_and_move_ptr(&body->local_iseq);
+        if (body->parent_iseq) rb_gc_mark_and_move_ptr(&body->parent_iseq);
+        if (body->mandatory_only_iseq) rb_gc_mark_and_move_ptr(&body->mandatory_only_iseq);
 
         if (body->call_data) {
-            struct rb_call_data *cds = (struct rb_call_data *)body->call_data;
-            for (unsigned int i=0; i<body->ci_size; i++) {
-                const struct rb_callinfo *ci = cds[i].ci;
-                const struct rb_callcache *cc = cds[i].cc;
+            for (unsigned int i = 0; i < body->ci_size; i++) {
+                struct rb_call_data *cds = body->call_data;
 
-                if (vm_ci_markable(ci)) {
-                    rb_gc_mark_movable((VALUE)ci);
+                if (reference_updating) {
+                    if (!SPECIAL_CONST_P((VALUE)cds[i].ci)) {
+                        rb_gc_mark_and_move_ptr(&cds[i].ci);
+                    }
+                }
+                else {
+                    if (vm_ci_markable(cds[i].ci)) {
+                        rb_gc_mark_and_move_ptr(&cds[i].ci);
+                    }
```

Comment:
Both branches have `rb_gc_mark_and_move_ptr(&cds[i].ci);`, which indicates the field is the same.  The difference is the condition of whether that field needs to be scanned.  I am not familiar with details about iseq, but I think iseq interacts with GC in some interesting way that this field `&cds[i].ci` may sometimes contain non-reference values that shouldn't be scanned by the GC.

From the code, it looks like the `ci` field may sometimes point to a struct allocated on the stack.  So I think this part of the code can be generalised to: *"if `&cds[i].ci` points to something on the stack, don't visit it"*.  This can be done without knowing whether we are in the marking phase or forwarding phase (which doesn't make sense for evacuating GCs anyway).

Reply:
I took a look a look at this, it looks like `ci` in the iseq can only be a Ruby object or null, and can't be stack allocated. I opened #7147.

Reply:
I see there is a macro `VM_CI_ON_STACK` and it defines a `rb_callinfo` with `VM_CALLINFO_NOT_UNDER_GC`.  Do you mean `iseq` can never point to such an `rb_callinfo` created by `VM_CI_ON_STACK`?  If that's true, I think this part can be simplified.

Reply:
Yes. I'll merge #7147 first and then update this code correspondingly.